### ISSUE TITLE
PKCE Code Verification Failure on Authentication

### ISSUE THREAD

**Issue by nickerzb** [okta/okta-oidc-js#804]
## I'm submitting this issue for the package(s):

- [ ] jwt-verifier
- [ ] okta-angular
- [ ] oidc-middleware
- [x] okta-react
- [ ] okta-react-native
- [ ] okta-vue

## I'm submitting a:

- [x] Bug report  
- [ ] Feature request
- [ ] Other (Describe below)

## Current behavior

When we initially login a brand new session, the user is able to authenticate successfully.  When we come back to our application (after a session has expired), the user is prompted to reauthenticate.  After entering credentials, the user is returned to our application, and sees "AuthApiError" and the network response is the following:
```{"error":"invalid_grant","error_description":"PKCE verification failed."}```

It seems like something is holding onto the old session for the PKCE flow when it triggers the login step.  We are able to get users into the application by refreshing the page, and triggering the authentication process again, but it's a bad user experience.

## Expected behavior

User is able to reauthenticate successfully on expired session.

## Minimal reproduction of the problem with instructions

I haven't been able to reproduce it every single time locally, but I set my token to expire in 5 minutes, wait for my session to end, then refresh the page.  There has to be a race condition somewhere.

## Extra information about the use case/user story you are trying to implement

Code we have implemented:

![image](https://user-images.githubusercontent.com/19597570/84273335-29e77c00-aaf4-11ea-8a68-1f65d9efe206.png)

![image](https://user-images.githubusercontent.com/19597570/84273469-4e435880-aaf4-11ea-816b-c22debd2eab8.png)

We are using the `/code=:code` path since we are forced to use a HashRouter with our current setup.  We also have a block of code directly using `@okta/okta-auth-js` that listens for unauthorized status codes in response from our API.

![image](https://user-images.githubusercontent.com/19597570/84273430-45eb1d80-aaf4-11ea-9fe0-5f2f0776f3b6.png)

Initially this was just calling the getWithRedirect function, but we added some code to try to clear out the user session before redirecting.

## Environment

- Package Version: 
  - @okta/okta-react -> 3.0.1 
  - @okta/okta-auth-js -> 3.1.1
- Browser: Chrome 83.0.4103.61
- OS: MacOs Catalina 10.15.4
- Node version (`node -v`): v13.9.0
- Other:

**Comment by restfulhead** [comment]
We are experiencing the same issue. For some users the PKCE verification sometimes fails and the request to the `/token` endpoint fails with a 400 error:

![okta-pkce-issue](https://user-images.githubusercontent.com/558375/84964289-be944000-b0d9-11ea-8f7a-79a471c2ac76.png)

My environment is a little different, but it sounds like the same issue.
```
"@okta/okta-angular": "~2.0.0",
"@okta/okta-signin-widget": "~4.1.1"
```

**Comment by andrewdmoreno** [comment]
I have been experiencing behavior similar to this while using `@okta/okta-vue 1.3.0` and just tried upgrading to `@okta/okta-vue 2.0.0` which uses updated `@okta/okta-auth-js` bits and I still encounter this from time to time. Seems like the only thing that works is manually clearing the okta-oauth cookies and refreshing the page. Has been very frustrating trying to track down exactly what is causing this as I have not found a sure fire way to reproduce it.

**Comment by nickerzb** [comment]
Yeah, this is intermittent for us as well.  Temporarily switched to the implicit grant within our application to avoid this.

**Comment by santirom** [comment]
I'm experiencing same issue using okta-vue. Sometimes I get 
`{error: "invalid_grant", error_description: "PKCE verification failed."}
error: "invalid_grant"
error_description: "PKCE verification failed."`

But if I try to login again it works

**Comment by aarongranick-okta** [comment]
@nickerzb @andrewdmoreno @santirom Thanks for reporting this issue. We have been able to identify some potential problems related to concurrency (or "race condition") with PKCE flow. Within an app this can happen if an async method is called without the `await` keyword or waiting for the promise to resolve. If concurrency is the issue, upgrading to the latest version of `okta-auth-js` (currently version 3.1.4) may help. This version includes an update which places all token API calls into a queue so that they cannot execute concurrently. If you are using one of our SDKs such as `okta-react` or `okta-vue`, wiping your node_modules and reinstalling dependencies should install the latest version of `okta-auth-js`

Another potential cause of this error is if the PKCE code is no longer valid. The timeout on the code is 60 seconds, so if a login has been sitting for over a minute it may fail with this error. Currently there is a default implementation of `onSessionExpired` which will bring the user to the login page. If this has been sitting for over a minute, the login may fail with this error. You can provide a custom implementation of `onSessionExpired`: https://github.com/okta/okta-auth-js#additional-options this function can be added to the same config object you are passing to construct an instance of whichever SDK you are using.

You may want to render a dialog telling the user that their session has expired and provide a link which will begin the login flow when clicked.

Please let us know if either of these suggestions was able to resolve your issue. Thanks!

**Comment by sleepwalky** [comment]
@aarongranick-okta 
I have a reliable way of reproducing this issue, not sure if it's correct though.

A bit of context:
"@okta/okta-auth-js": "^3.1.4",
"@okta/okta-react": "^3.0.2",

I'm using @okta-react and SecureRoute and I extracted Auth service into a separate file. Auto renew of token is enabled by default.
```
import { AuthService } from '@okta/okta-react'
import { registry } from 'routes/registry'
import { config } from 'config'

const CLIENT_ID = config.OKTA_CLIENT_ID || ''
const ISSUER = config.OKTA_ISSUER || ''

export const oktaConfig = {
  clientId: CLIENT_ID,
  issuer: ISSUER,
  redirectUri: `${window.location.origin}${registry.loginCallback}`,
  scopes: ['openid', 'profile', 'email'],
  pkce: true,
  disableHttpsCheck: false,
}

export const Auth = new AuthService(oktaConfig)
```

My App.tsx looks almost like that:
```
import React, { FC } from 'react'
import { Router, Switch, Route } from 'react-router-dom'
import { Security, SecureRoute } from '@okta/okta-react'
import { Provider } from 'react-redux'

import store from 'redux/store'
import {
  HomePage,
  GroupsPage,
  ProfilePage,
} from 'pages'
import { Bootstrap } from 'components/Bootstrap'
import { LoginCallback } from 'components/LoginCallback'
import { registry } from 'routes/registry'
import { Auth, history } from './services'

const App: FC = () => {
  return (
    <Router history={history}>
      <Security authService={Auth}>
        <Provider store={store}>
          <Bootstrap />
          <Switch>
            <Route path={registry.loginCallback} component={LoginCallback} exact />
            <SecureRoute path={registry.groups} component={GroupsPage} exact />
            <SecureRoute path={registry.group(':groupId')} component={GroupDetailsPage} exact />
            <SecureRoute path={registry.profile} component={ProfilePage} exact />
            <SecureRoute path={registry.home} component={HomePage} />
          </Switch>
        </Provider>
      </Security>
    </Router>
  )
}

export default App
```
In every http call to our BE I retrieve token and add it to http-headers:
```
  const token = await Auth.getAccessToken()
  // sometimes token is undefined, so I read somewhere in the docs that re-login is required
  if (!token) {
    Auth.login(window.location.pathname)
  }
```

Before Switch component, I'm loading some data in Bootstrap component:
```
export const Bootstrap: FC = () => {
  const { authState } = useOktaAuth()
  const dispatch = useDispatch()

  const { isAuthenticated } = authState

  useEffect(() => {
    // added this extra check in hope that it won't trigger renew of token logic when users lands on loginCallback url
    if (!isAuthenticated) return
    initializeFirebase()
    dispatch(getLoggedInUser.thunk())
    dispatch(getPermissions.thunk())
  }, [isAuthenticated])

  return null
}
```

So, speaking about how I can reproduce the issue:
- I manually delete okta-token-storage from localStorage
- then I click on one of my links which are usually head user to SecureRoute
- some of the routes are handled correctly - I see re-login and redirect to loginCallback 
- but for other links I get "PKCE verification failed" error in 100% of attempts.

If my way of reproducing the error is okay, I'm ready to help you guys to fix the issue.

Also please correct me if my implementation has mistakes in terms of how okta should be used.

**Comment by aarongranick-okta** [comment]
@sleepwalky Thank you for providing a detailed reproduction. I think the problem may be caused by beginning a new authorization flow while processing a login callback. This could be caused by an explicit call to `login` or by the combination of `autoRenew` and calling `getAccessToken` on the callback route. 

This code: https://github.com/okta/okta-auth-js/blob/master/packages/okta-auth-js/lib/TokenManager.js#L123 will attempt to renew a token, which begins an authorization flow and writes new PKCE data to session storage. If this occurs on the callback route, it will overwrite the PKCE data which was saved before the redirect. If you see a network call to `/authorize` with a query parameter `response_mode=okta_post_message` that is most likely the token renew.

All of our samples process callback on a dedicated route and take care not to perform any other logic on the callback route. Attempting to read the access token or determine authentication status or any other API which may begin a a new authorization flow while on the callback route may cause this unpredictable behavior.

For example, if this code runs on the callback route, you will definitely see an issue:

```
  const token = await Auth.getAccessToken()
  // sometimes token is undefined, so I read somewhere in the docs that re-login is required
  if (!token) {
    Auth.login(window.location.pathname)
  }
```
The access token will not exist because it has not been retrieved and stored yet. The call to `login` will overwrite the PKCE data. This will most likely cause an error somewhere in the login callback, depending on timing. (it's a race/concurrency condition).

Also be aware that `login` is an async function, you should `await` it. Even if it is going to perform a redirect it must calculate the PKCE code first, which is an asynchronous operation. It is possible that while this calculation is happening - but before the redirect - some other code may run, see that the token is missing/invalid, and begin another authorization flow which corrupts the data.

**Comment by santirom** [comment]
I fixed my issue. I found a call to async method without await, as @aarongranick-okta said.
Also I updated okta-signin-widget to latest version

Thanks for the help!

**Comment by andrewdmoreno** [comment]
> @sleepwalky Thank you for providing a detailed reproduction. I think the problem may be caused by beginning a new authorization flow while processing a login callback. This could be caused by an explicit call to `login` or by the combination of `autoRenew` and calling `getAccessToken` on the callback route.

@aarongranick-okta  Based on your explanation, I _think_ I may have found what was causing my issue but wanted to get your thoughts to confirm. I have data that I fetch from the server that is needed pretty much on all routes of my app. Fetching this data requires a token. While I do not have any other auth related calls specifically in my callback route - I do, however, have a `router.beforeEach` that was making an async call to my server if user was authenticated and config had never been fetched before:

````
router.beforeEach(async (to, from, next) => {
  if (await Vue.prototype.$auth.isAuthenticated() && store.state.initCompleted === false) {
    // Perform initialization
    await store.dispatch(FETCH_CONFIG);
    store.commit(INIT_COMPLETED);
  }
  return next();
});
````

In its current state this code would also be running before `ImplicitCallback` as it is in a `router.beforeEach()`. Would this cause some of the conflicts you mentioned above? If this was likely the culprit I was considering updating the above code to specify that the initialization code should only be invoked on routes that do not have `meta.requiresAuth = true`:

````
to.matched.some((record) => !!record.meta.requiresAuth === true)
````

Would this work or do you have any other suggestions?

**Comment by aarongranick-okta** [comment]
@andrewdmoreno Yes, calling `isAuthenticated` or `getUser` on the callback route can cause the issue. Your idea of running the logic only on routes which require authentication seems like a good strategy to avoid making these calls on the callback route.

**Comment by aarongranick-okta** [comment]
internal ref: OKTA-306578

**Comment by nickerzb** [comment]
The updates to the library have not resolved my issues.  Still getting `{"error":"invalid_grant","error_description":"PKCE verification failed."}`

**Comment by santirom** [comment]
@aarongranick-okta 
I'm experiencing the issue again in a particular way. If the user goes back to the website after some time, then it's redirectered to okta login, the difference is that the url is https://dev-XXXX.okta.com/login/login.htm?fromURI=/oauth2/v1/authorize/redirect?okta_key=XXXXXXXXXXXXXXXXXXXXXXXXX instead of myurl.com/login, so after authentication the callback url is called but I get the PKCE verification failed. Any ideas of what could be causing the issue?

**Comment by sleepwalky** [comment]
@aarongranick-okta 
Alright, I've refactored my solution to not trigger any calls on the callback route. Also, I've deleted the call to `login` in my http utility, so it won't do anything bad when I'm making 2 requests in parallel. However, a couple of questions arised: 
1) what should I do if `Auth.getAccessToken()` returns undefined?
2) what should I do if backend returns 401 Not Authorized response?

**Comment by igcherkaev** [comment]
Pretty much same problem using Angular and Okta Auth component. Very annoying and bad UX.

**Comment by myspivey** [comment]
This looks to be occurring in the React component as well.

**Comment by myspivey** [comment]
This appears to be the same problem we are having with React. As I understand it, this is an issue of calling `authService.getUser()` before the re-auth flow is complete. This was resolved for others by ensuring that we do not call this during the callback flow. The issue we have is that we are not doing this through the callback flow. We grab `authState` from `useOktaAuth()` and watch that with a `useEffect` hook. When we see the state update to `isAuthenticated`, we call getUser to do some other things. I am unsure considering we are not dealing in callbacks the best solution to resolve this. If this is a timing issue, it might be that we need to have a timeout but that feels wrong. I am curious if the SDK should be instead waiting to set `isAuthenticated` until the flow is fully complete.

**Comment by aarongranick-okta** [comment]
@nickerzb @sleepwalky @igcherkaev @myspivey 
If the issue is related to session expiration / token renew failure, it may be related to the default implementation of [onSessionExpired](https://github.com/okta/okta-auth-js#additional-options). Currently, this is implemented in Angular/React SDKs to automatically begin a new login flow. This will redirect user to the Okta login page. If the user does not complete the flow within 60 seconds, it may fail. For this reason we are planning on changing or removing this default implementation. At this time, we recommend passing your own function for `onSessionExpired` and calling `logout`. This will clear all the invalid tokens from storage. You may want to notify the user that their session has expired before doing this step, since it will return the app to the unauthenticated state.

@sleepwalky If backend returns 401 this means the user session does not exist or has expired. You can start a new login flow. If `getAccessToken` returns undefined it means that there is no token in storage or the token in storage was invalid and could not be renewed. In either case, you may start a new login flow to get a new token. If a session for the user does not exist they will be prompted for credentials, otherwise the redirect will return to your app immediately with the code for token exchange.

@myspivey If your component is active on the callback route, it could be related to this issue. There should be NO other logic which tests authentication state on the callback. When the callback is complete, it typically redirects to another (the original) route. Your component with effectHook should only be active on this other route. If it is being activated on the callback route, there may be a race condition. The other cases users are reporting here are related to session expiration / token renew failure. If your issue is not related to or login callback or session expiration / token renew failure , please open a new issue so we can start a fresh conversation. Also be aware that it is possible for `isAuthenticated` to return a false positive, since it tests only for the *presence* of the idToken and accessToken. If the session was closed or expired but the tokens remain in storage, this method may return true. We provide an option for customizing this behavior: https://github.com/okta/okta-oidc-js/tree/master/packages/okta-react#configuration-options Calling `getUser` is also a way of testing if the user's session is valid.

**Comment by igcherkaev** [comment]
@aarongranick-okta it doesn't help. I redirect users to /logout on session expire and it seems to be working when fairly short time passed since user interacted with the site, however, if I keep browser open for longer (several hours) but idling, and it goes via the same flow, it fails even though it goes via oktaAuth.logout(). It definitely completes the flow within 60 seconds.

Is there any way I can handle errors happening when `/implicit/callback` is called by OktaAuth?

```
{ path: 'implicit/callback', component: OktaCallbackComponent },
```

Any way I can tap in and maybe redirect users again to /login if any error occurs?

**Comment by polklabs** [comment]
@igcherkaev One way i found to handle the error for Angular while I look for a real solution is to copy the OktaCallbackComponent and then add the following code to the .catch() statement
```
if (e.xhr.responseText === '{"error":"invalid_grant","error_description":"PKCE verification failed."}') {
     this.router.navigateByUrl('/login');
}
```

They way my login component is set up, it tries to login the user automatically and if it cant it'll show the login widget. When the PKCE error occurs it successfully logs the user in to Okta so redirecting back to the login route automatically gets the token without an error.

**Comment by igcherkaev** [comment]
@polklabs
Thanks! I am not a pro in Angular, but is there a better way than copy the component? Maybe extend it with your own class and override some method?

**Comment by polklabs** [comment]
@aarongranick-okta I have noticed that when I get the pkce code error, the authorize and token endpoints are using wrong/mismatched values.

/authorize is passed **AAAAAA**  for the code_challenge
/token is passed **BBBBBB** for the code_verifier

But calling `computeChallenge()` from okta-auth-js/lib/pkce.js where
```
computeChallenge(BBBBBB) != AAAAAA
```

Alternatively the codeVerifier from okta-pkce-storage in local storage is **CCCCCC** and does map to the passed code_challenge but it is not being used.

Could there be something in the okta-signin-widget, okta-oidc-js, or okta-auth-js that is calling generateVerifier twice when under certain conditions like the Okta session and token being expired?

When I log out normally and log back in, the okta-pkce-storage codeVerifier matches the value passed in /token.
I am not calling any of these endpoints myself so none of my code should be affecting them

**Comment by metalsho** [comment]
Same problem here with angular and okta widget. Right now I use a custom callback component, but I have to redirect the user to login page and ask him to try again.

**Comment by myspivey** [comment]
As a follow up, we have been working on this now for a week with a lot of mixed luck that still mostly ends in errors. We had gotten feedback that our issue was that we declared our own AuthService even though the docs say this is supported. We ended up removing that and using the "built in" one to avoid conflicts which did not end up solving the issue. The next feedback we got was that we were calling logout on `onAuthRequired` and `onSessionExpired` and instead should only call it when seeing events off `authService.getTokenManager`. This is the fix I tried last night using an effect to capture the error, looking at the error code if it is `login_required` and then sending the user to the login. This appears to work at first as it takes me to login. It does throw an error in the console which is rather frustrating since we are handling the error. The broken part is that when you then try to login, you get a 400 error calling `/oauth2/default/v1/token` and it simply displays the generic `Error: AuthApiError` on the screen. So we are still basically at square one here where we have managed to get around the PKCE error only to still be stuck on an error.

**Comment by restfulhead** [comment]
We're also still facing the same issue. We've removed all custom calls in the callback handler and now also always force a logout when the token expires (as suggested in some of the comments). It did not help and users are still seeing the issue occasionally. The second login attempt then always works, but this is a bad user experience. I'm not sure what else to do...

**Comment by aarongranick-okta** [comment]
@igcherkaev We recommend copying the callback component and adding custom logic for error handling. There is no relevant logic except calling `handleAuthentication()` on the callback route. Your component can add additional handling for errors including showing the users a friendly message or link back to the login flow. https://github.com/okta/okta-oidc-js/tree/master/packages/okta-angular#oktacallbackcomponent

**Comment by aarongranick-okta** [comment]
@polklabs The error you are seeing (where the code in storage does not match the request) is a symptom of concurrency. There is more than one call to `generateVerifier` before the flow has completed. As mentioned in previous comments, this can happen if your AuthService is performing logic on the constructor. It may also happen if there is logic which normally executes on a secure route is being performed on the callback route (the callback is an unauthenticated route) or if there is logic which tries to test if the user is logged in. You mentioned the code is in *local* storage. The latest version of `okta-auth-js` should be using *session* storage to avoid concurrency issues when multiple tabs are open. Please ensure you have the latest version installed underneath `okta-angular/node_modules`. If you wipe your node_modules directory and re-install it should install the latest version automatically.

@myspivey @restfulhead 
If possible, try to detect the source of the concurrency using the debugger. The effects of the concurrent calls could also be mitigated by setting `autoRenew` to false. You will need to listen for the 'expired' event from TokenManager and renew tokens manually using `TokenManager.renew`. Note that `onSessionExpired` will not be called if `autoRenew` is turned off. (`onSessionExpired` is called when token auto renew fails). 

If you are certain that concurrency is not the issue and turning off `autoRenew` does not mitigate the problem, please create a separate issue so we can dig deeper into the possible source of the issue. Thanks!

**Comment by polklabs** [comment]
@aarongranick-okta 
My login routes (/implicit/callback, /login, /logout) do not use oktaAuthGuard so are not secure.

Only two places call `this.oktaAuth.isAuthenticated();` and one calls `this.oktaAuth.getIdToken()` none of which are being hit between hitting the login button and when i get the error. I'm skipping the `isAuthenticated()` calls if the route is "/implicit/callback".

Only a single call to each authn, authorize and token endpoints are being made. 

Using Okta versions:
```
"@okta/okta-angular": "^2.1.0",
"@okta/okta-auth-js": "^3.2.1",
"@okta/okta-signin-widget": "^4.2.1",
```

It looks like the okta-pkce-storage in session storage is being filled in and cleared out during login, didn't notice before since the oktaAuthService was clearing out the data before i could see it, but the okta-pkce-storage in localStorage is also being filled in. Both session and local storage values match when `handleAuthentication()` in the callback component is called but the code_verifier passed into the /token call is different.

Error only occurs when the Okta session (not token) has expired and works perfectly every other time.

**Comment by aarongranick-okta** [comment]
@polklabs If you are seeing PKCE codes stored in *both* local storage and session storage, my suspicion is that the issue is coming from `okta-signin-widget` using an older version of `okta-auth-js`. Can you compare the version of `okta-auth-js` in `@okta/okta-signin-widget/node_modules` with the version in `@okta/okta-angular/node_modules`? On a clean install of node_modules, the widget should use the latest version, based on the semver `^3.1.1`. `3.1.2` is the version that changed PKCE storage to session storage: https://github.com/okta/okta-auth-js/releases/tag/okta-auth-js-3.1.2

**Comment by polklabs** [comment]
@aarongranick-okta Originally the `okta-auth-js` in my node modules was `3.1.4`. I just did a npm install of version `3.2.1`. Our publishing system runs a fresh `npm install` before each build. 
So now my package.json file has the versions stated in my previous comment. And i deleted my `node_modules` folder and did a npm install.
Both locally and on our staging machine the local storage `okta-pkce-storage` is being set.

**Comment by restfulhead** [comment]
@aarongranick-okta I'm certain that our callback component does not perform any other calls. We do have our own component, but the logic is very similar to the out of the box [OktaCallbackComponent](https://github.com/okta/okta-oidc-js/blob/master/packages/okta-angular/src/okta/components/callback.component.ts). The only additionally logic is catching the PKCE error similar to what @polklabs posted a few messages above. Then the user has to enter their credentials again and the second time it works.

I've updated the libraries again today and still saw it occur once. My env:
```
@okta/okta-angular: 2.1.0
@okta/okta-auth-js: 3.2.1
@okta/okta-signin-widget: 4.2.3
```

I've noticed something else, which may be related. After I successfully log in, I see an empty `okta-pkce-storage` entry in the session storage. But: I also see a populated `okta-pkce-storage` in local storage. And: This does not get cleared out even when logging out. Could it be that certain fall-back logic (such as [this](https://github.com/okta/okta-auth-js/blob/master/packages/okta-auth-js/lib/pkce.js#L61)) may interfere here and causing the wrong verifier value to be used?

**Comment by igcherkaev** [comment]
I can also confirm that my app also has okta-pkce-storage with codeVerifier in it in the local storage. Using latest modules and sign in widget.

**Comment by abodnar63** [comment]
We have similar errors in React App. 
We use Okta Authorization Code Flow with PKCE.
Packages:
```
@okta/okta-react: "3.0.2"
@okta/okta-auth-js: "3.2.1"
```
Callback URL from Okta handles `LoginCallback` component: `import { LoginCallback } from '@okta/okta-react';`
When user tries to sign in for the first time he could get different errors:
```
PKCE verification failed
Could not load PKCE code verifier from storage
token endpoint 400 error {"error":"invalid_grant","error_description":"PKCE verification failed."}
```
Refreshing the page with callback signs user in without any errors. Then login/logout work for a user for the next 4-6 hours and this error appears again.

Currently, our temporary solution is to catch errors like that and refresh the page only for the first time to avoid cycling.

Would be great to know how we can avoid these errors.

**Comment by aarongranick-okta** [comment]
@abodnar63 @igcherkaev @restfulhead @polklabs @metalsho @myspivey @sleepwalky and anyone else experiencing this issue:

Can you confirm if you are "self hosting" the signin widget? Are you including `@okta/okta-signin-widget` in your package.json and rendering it on a page within your application? Or are you redirecting to Okta for login?

**Comment by sleepwalky** [comment]
@aarongranick-okta I'm not using signin-widget

**Comment by polklabs** [comment]
@aarongranick-okta We are self hosting the widget. We render it in our application

**Comment by myspivey** [comment]
@aarongranick-okta self hosted

**Comment by calvinmuller-wi** [comment]
self-hosted custom login page.

**Comment by abodnar63** [comment]
@aarongranick-okta, We use Okta authentication flow and do not have signin-widget

**Comment by aarongranick-okta** [comment]
Thank you for your responses. I've created a new internal issue to track this (OKTA-318184) with high priority. We will work on reproducing and fixing this issue and will provide updated information here.

**Comment by igcherkaev** [comment]
@aarongranick-okta self-hosted custom login page.

**Comment by amcdnl** [comment]
@aarongranick-okta - self hosted. Whats the update - my app is almost completely broken as it stands right now.

**Comment by shuowu** [comment]
@amcdnl yes, it's relevant. we suspect the default auto login behavior in onSessionExpired caused the concurrency issue for PKCE. So the fix should be just clearAuth state in react without auto login to log users out.

Also, I remember I took a look of your gist last week, you probably can try to remove related window.location.replace('login') in code to avoid this error.

**Comment by aarongranick-okta** [comment]
@abodnar63 @igcherkaev @restfulhead @polklabs @metalsho @myspivey @sleepwalky @amcdnl Thanks for your responses. We are currently working to fully understand and fix this issue. We think it may be related to the default behavior for `onSessionExpired` which may be causing a concurrency error. As a workaround, we suggest adding a custom implementation of `onSessionExpired` to your config which can be an empty function:

```
var config = {
  issuer: ....
  clientId: ....
  onSessionExpired: () => {}
};
```
We do not believe that this will cause any negative side-effects, since the `onSessionExpired` default behavior is currently NOT working as it was intended. But we think there is a good chance it may resolve this concurrency-related error with the PKCE storage.

If you have the time and ability to test this workaround, we would be grateful for any feedback.

**Comment by amcdnl** [comment]
@aarongranick-okta - This is what I have but its not working:

```
<Security
  {...authconfig}
  onAuthRequired={() => history.push('/login')}
  onSessionExpired={() => history.push('/logout')}
>
```

**Comment by aarongranick-okta** [comment]
@amcdnl @polklabs The problem we have identified is that `onSessionExpired` may be called as part of the normal auth flow triggered from a SecureRoute. Calling either `login` or `logout` while in the middle of another auth flow may cause this concurrency error. This is why we are recommending to override with an empty function at this time.

**Comment by aarongranick-okta** [comment]
@amcdnl @igcherkaev @calvinmuller-wi @myspivey @polklabs

I was able to reproduce the issue using the self-hosted widget. The cause of the issue is:

- current version of the widget stores PKCE code in localStorage
- current version of authJS looks for code in sessionStorage, but falls back to localStorage if code is not present
- token renew failure does not cleanup but leaves a code in sessionStorage. token renew failure can occur if the session is expired OR if 3rd party cookies are blocked 
- the error occurs because the code from token renew (in sessionStorage) is passed to the `/token` endpoint instead of the code saved from the widget (in localStorage)

I was able to workaround this issue by executing this code right before rendering the widget:

```
sessionStorage.setItem('okta-pkce-storage', '{}')
```

This will clear out any old PKCE codes from sessionStorage before beginning the login flow. If you add this in your login component it should fix the issue you are seeing.

 We will work on getting a permanent fix implemented and released in the very near future so this workaround will not be necessary (although it should not cause any negative effects. The widget is the beginning of a PKCE auth flow so there should be no valid data saved in sessionStorage)

Thank you for your patience while we investigated this issue.

**Comment by corchard** [comment]
Seeing the same issue as @abodnar63 also using okta-react without the sign in widget.

`AuthSDKError: Could not load PKCE code verifier from storage`

**Comment by aarongranick-okta** [comment]
@corchard @abodnar63 @sleepwalky (and anyone else not using signin widget)

I believe you are seeing a different issue than the one we have diagnosed here with a custom hosted signin widget.

The error `AuthSDKError: Could not load PKCE code verifier from storage` is possibly related to a concurrency error. It could mean that an auth flow has completed and cleared storage before the callback has occurred.

This error could also occur if there is an attempt to handle an auth callback more than once. When the callback is handled (success or failure), the PKCE storage will be cleared. A second attempt to handle the callback will hit this error.  We have seen this error when more than one callback component is rendered on the callback route.

This error is also distinct from the error we are seeing on this issue: `token endpoint 400 error {"error":"invalid_grant","error_description":"PKCE verification failed."}` which is related to the *wrong* code verifier in storage. This again, could be caused by concurrency - multiple auth flows running in parallel. Most commonly caused by failing to *await* an *async* method.

Hopefully this helps find the source of the problem. If you are sure you don't have multiple callback handlers and are continuing to see an issue with the error `AuthSDKError: Could not load PKCE code verifier from storage`, please create a separate Github issue and try to provide as much relevant information as you can.

If you have eliminated concurrent auth flows as a possibility and are still seeing `token endpoint 400 error {"error":"invalid_grant","error_description":"PKCE verification failed."}` *without* a signin-widget in your application, please create a separate Github issue and (as usual) try to provide as much relevant information as you can.

We understand these errors are causing disruptions in your apps. We are committed to finding the causes of these errors and, if caused by a bug in an Okta SDK, fixing the issue quickly. We will also consider ways that the SDK can provide better error messages to help diagnose these issues.

**Comment by amcdnl** [comment]
On a side note: I'm getting a `Cannot read property 'message' of undefined` everytime I do login correctly.

![image](https://user-images.githubusercontent.com/227909/89055673-cfff7780-d328-11ea-9a85-5ebfeba12c18.png)

**Comment by abodnar63** [comment]
> @amcdnl @polklabs The problem we have identified is that `onSessionExpired` may be called as part of the normal auth flow triggered from a SecureRoute. Calling either `login` or `logout` while in the middle of another auth flow may cause this concurrency error. This is why we are recommending to override with an empty function at this time.

@aakashyadav-okta  Overriding of `onSessionExpired` helped to solve my case. Instead of the empty function, we need to redirect the user to the Okta login page otherwise when the session is expired we are getting an error like: `session is expired and the user should be prompted to login`.

**Comment by amcdnl** [comment]
If I keep my browser open for extended amounts of time, it still throws this error:

![image](https://user-images.githubusercontent.com/227909/89065958-9c2d4d80-d33a-11ea-9895-10126561b093.png)

**Comment by aarongranick-okta** [comment]
@abodnar63 The console message `error.errorCode: login_required, error.description: The client specified not to prompt, but the user is not logged in.` is expected when token renew fails. Token renew can fail if the session has expired, or if 3rd party cookies are blocked. In either case, token renew failure will remove the invalid accessToken from storage and call `onSessionExpired` handler. 

Token renew is probably being triggered when the user is navigating and `isAuthenticated` is being evaluated on a SecureRoute. An alternative to implementing `onSessionExpired` is to provide a custom implementation for `isAuthenticated`. The default implementation of `isAuthenticated` returns true if EITHER accessToken or idToken exists. You may get better results testing for ONLY the accessToken or for BOTH accessToken and idToken. https://github.com/okta/okta-oidc-js/tree/master/packages/okta-react#configuration-options

**Comment by amcdnl** [comment]
Any updates?

**Comment by aarongranick-okta** [comment]
@amcdnl We have released a new version of `@okta/okta-auth-js` (3.2.2) which should correct this flaw. The latest version of `@okta/okta-auth-js` will be installed as a dependency by our React or Angular SDK on a clean install. To update an existing install, we recommend erasing your node_modules directory and any .lock files and then running `npm i` or `yarn install` (if you are using yarn).

**Comment by igcherkaev** [comment]
@aarongranick-okta I am testing this with Angular 9 and:
- `@okta/okta-auth-js: 3.2.2`
- `@okta/okta-angular: 2.1.0` 
- `@okta/okta-signin-widget: 4.2.2`

Seems like latest versions of `@okta/okta-signin-widget` has some bugs rendering the widget unusable in some cases.
So far looks good, but I'll give it more tries and report back.

**Comment by aarongranick-okta** [comment]
@igcherkaev Is the render bug with widget version 4.2.2, or with newer versions?

**Comment by igcherkaev** [comment]
@aarongranick-okta with the newer versiond (4.2.3 and 4.3.0), I left a note in the existing issue there in the widget repo.

**Comment by igcherkaev** [comment]
@aarongranick-okta `@okta/okta-auth-js: 3.2.2` seems to be working fine now, no PKCE issues so far.

**Comment by ggentzel** [comment]
I am seeing this same error when using `@okta/okta-vue: 2.0.0`, I have confirmed that the latest version `@okta/okta-auth-js: 3.2.2` is being used. It doesn't always happen but one of our users reports he can make it happen consistently. I have a video of the issue with the user clicking through his developer tools in chrome that I can provide if that is of help. Any suggestions?

**Comment by aarongranick-okta** [comment]
@ggentzel If you are able to post that video here, that could be helpful.

**Comment by myspivey** [comment]
Still getting issues in this space it seems. I am curious if in the short term there are at least ways to capture some of these issues rather than chucking errors? We get a fair amount of reporting to Sentry and locally with this issue that creates a lot of noise that seems to not be actionable yet. Stuff like this. 
![image](https://user-images.githubusercontent.com/2472925/92329575-77a54f00-f025-11ea-815a-a0369864c40b.png)

**Comment by aarongranick-okta** [comment]
@myspivey This could be caused by the `/callback` route loaded twice. The PKCE code is one-time use and is deleted after it is used. A second call to `/callback` would produce an error like this. In React this can happen based on route config. We recommend using the `<Switch>` component and setting the `exact` keyword on the callback route. Another possibility is multiple concurrent auth flows, which could occur whenever an asynchronous function (like loginRedirect) is called without the `await` keyword. This includes in event handlers or callback functions like `onAuthRequired` or `onSessionExpired`

**Comment by dlauterbach** [comment]
I'm using React (with @okta/okta-react/3.0.7 and okta-auth-js/3.2.4) and was experiencing both issues:  `{"error":"invalid_grant","error_description":"PKCE verification failed."}` and `AuthSDKError: Could not load PKCE code verifier from storage.`  I, like @nickerzb, set my token to expire every 5 minutes to speed up debugging.  I would receive `AuthSDKError: Could not load PKCE code verifier from storage.` after 5 minutes of leaving my application open in my browser as result of attempted token renew.  For me the fix was to remove my node_modules folder and running `npm install` as @aarongranick-okta advised.  I may have previously used `npm update` to upgrade to these latest okta packages, not sure.  A complete re-install of all npm packages seems to have resolved my issues.

**Comment by aarongranick-okta** [comment]
@dlauterbach We are glad you were able to resolve your issue! okta-react has an internal dependency on okta-auth-js and even if your app depends on okta-auth-js, another copy is usually installed in `node_modules/okta-react/node_modules/@okta/okta-auth-js` It's likely that this copy needed to be updated, and a clean install will do that.

**Comment by aarongranick-okta** [comment]
@nickerzb 

I believe this (intermittent) issue was related to the "passive" auto renew implemented in previous versions of `okta-auth-js`. 

Attempting to renew tokens in the middle of an authentication flow would cause the PKCE metadata to be overwritten.

In version 4.0 we introduced a [new model for auto renew](https://github.com/okta/okta-auth-js/pull/444) and changed the behavior of `tokenManager.get` so that it would not trigger a token renew operation. 

Additionally we have added concurrency protection to all "token" operations so they will always execute sequentially. 
We believe that the issue described here has been resolved in the current version of `okta-auth-js`

For any and all issues with `okta-auth-js`, please visit https://github.com/okta/okta-auth-js/issues

**Comment by mpelham** [comment]
@aarongranick-okta, quick question:
> We recommend using the <Switch> component and setting the exact keyword on the callback route. Another possibility is multiple concurrent auth flows, 

If `exact` is recommended on the Login Callback route, why is this not part of the standard documentation?
